### ISSUE TITLE
Accept format 10x slower than URL format mystery 🔍

### ISSUE THREAD

**Issue by sfcgeorge** [rails/rails#30502]
### Behaviour & steps to reproduce

If you request a javascript view using the Accept header to specify the format, Rails seems to slow down partial rendering almost 10x. If you simply add .js to the URL (with or without Accept header) Rails does its thing at normal speed.

Output is the same, everything else is the same, no DB queries; just not having .js on a URL results in a 10x slowdown. 

Boom minimal repro with further details and logs: https://github.com/sfcgeorge/url-format-speed

### Expected behavior

The Accept javascript header / adding .js to the URL _should_ do the same thing right? Both tell Rails it's javascript format. The output and logs are the same in both cases, so what extra thing could Rails be doing for every partial rendered? Something something caching?

### System configuration
**Rails version**: 5.1.3
**Ruby version**: 2.4.0

**Comment by madejejej** [comment]
@sfcgeorge I could reproduce this in development environment but in production environment the difference is negligible. Could you confirm this is the case for you as well?

**Comment by sfcgeorge** [comment]
@madejejej Yes in development. In my real world app it's making certain actions take 2s when they should take ~300ms. I realise developmeant is "meant" to be slower but usually it's constant. It's important to be able to compare actions before shipping a feature to see "hmm this action is slow, is better add some indexes" or similar.

**Comment by K-S-A** [comment]
We've noticed such performance issues with .js.erb views (rails 4.2.5 and 5.1.0).

They took 300ms – 500ms to render first time in the production environment. All subsequent requests were processed "normally".

In development, all requests were processed with the next timings:
- first time render 300+ms;
- subsequent partial render - 3ms – 5ms.

NOTE: .js partials were rendered without issues in both production and development environments.

**Comment by madejejej** [comment]
On one hand I really think that if you want to know how something will behave on production you should have some kind of a pre-production environment which is configured the same as production.

On the other hand I agree that this is a strong outlier in development mode and can lead to confusion. 

I profiled the code with RubyProf and it looks like around 70% of time is spent inside `ActionView::PathResolver#find_template_paths` which calls `Dir#[]` with a complicated query like:

`url-format-speed/app/views/tests/_partial{.en,}{.js,.application/ecmascript,.application/x-ecmascript,.html,.text,.css,.ics,.csv,.vcf,.png,.jpeg,.gif,.bmp,.tiff,.svg,.mpeg,.xml,.rss,.atom,.yaml,.multipart_form,.url_encoded_form,.json,.pdf,.zip,.gzip,}{}{.raw,.erb,.html,.builder,.ruby,.coffee,.jbuilder,}`

In production environment `Dir#[]` is called only once and somehow sub

sequent calls are cached. I don't yet fully grasp the code and I'll try to dig into it more later to see if I can do something about it.

I'd appreciate some help from the Rails core team on this ❤️ 

Attaching the results from RubyProf. One profile called `prod-without-js` where I ran the curl command (after cache warmup) and the second one called `without-js` which comes from development mode.
[prof.zip](https://github.com/rails/rails/files/1273008/prof.zip)

**Comment by ahorek** [comment]
there's a known perf regression in ruby
https://bugs.ruby-lang.org/issues/13167

I was wondering if something can be done on the rails side, because these lookups are too complex. I don't think you have a file with .application/ecmascript extension in your application, but each possible format will slowdown the query twice!

**Comment by madejejej** [comment]
We probably could translate these mime types to possible extensions. However, as @sfcgeorge noticed the following remote call:

`link_to "I'm Sorry", podbay_doors_user_path(@dave), remote: true`

generates this HTTP Accept header:

`Accept: */*;q=0.5, text/javascript, application/javascript, application/ecmascript, application/x-ecmascript`

which forces Rails to search for all known formats. But there are use cases for that so I don't know if something can be done around this 😢 

The good thing is that the partials get cached so this isn't an issue in production (at least after the first render).

**Comment by matthewd** [comment]
It sounds like a bug (data error?) that we're looking for `.application/ecmascript` as an actual extension. But yeah, Accept headers are less "strong" than setting a `:format`, so it will lead to a broader query.

It's also true that development and production differ: production caches various things more aggressively, including partial lookups in this case. Development is obliged to notice if a file of one of those names appears at runtime, whereas production is allowed to assume that nothing changes after boot.

If the above path query is representative, though, it seems we could work faster by using the extra knowledge we have: as nobu notes on the ruby issue, `Dir[]` must contend with patterns with `/` inside a `{}` group, for example. If we know that's not the case for us, it seems likely to be faster for us to apply a much simpler pattern (`.../tests/_partial*`, for example), and then do the final filtering ourselves.

We do need to be careful here for security issues, particularly with regard to differing behaviours on case-sensitive vs -insensitive filesystems... but if anyone's interested, a PR to do the above sounds worth a go.

We could also consider allowing development to cache this better, and rely on a file watcher to clear that cache when files (dis)appear. That seems suited to a second, follow-on, PR... merging it with the above is likely to get confusing.

**Comment by madejejej** [comment]
@matthewd I started to dabble in the code to simplify the `Dir.glob` query. It's doable if we remove one unused feature in the `PathResolver` class: the ability to specify a custom pattern:

```ruby
  class PathResolver < Resolver #:nodoc:
    EXTENSIONS = { locale: ".", formats: ".", variants: "+", handlers: "." }
    DEFAULT_PATTERN = ":prefix/:action{.:locale,}{.:formats,}{+:variants,}{.:handlers,}"

    def initialize(pattern = nil)
      @pattern = pattern || DEFAULT_PATTERN
      super()
    end
  end
```

I grepped through the code looking for all descendants of `PathResolver` and `PathResolver` itself and it looks like the default pattern is overridden only in test files. 

Without removing this feature it gets much harder. For example in one of the test files the pattern looks like this:

```ruby
  def setup
    path = File.expand_path("../fixtures", __dir__)
    pattern = ":prefix/{:formats/,}:action{.:formats,}{+:variants,}{.:handlers,}"
    @resolver = ActionView::FileSystemResolver.new(path, pattern)
  end
```

My approach would be to change the `build_query` implementation to build a path without any locales, formats, variants nor handlers and then after finding all potential files with `Dir.glob` either:

- use regexes to filter out correct filenames, or;

- attempt to create all `Template` objects and filter out the ones that are not interesting to us

The second approach looks easier because there's already a method that extracts ` handler, format, variant` from the template filename. 

Does it all make sense? 😄

**Comment by matthewd** [comment]
The regex approach was my first thought too.. but can we do it with fnmatch? I think that'd allow us to retain the existing feature.

Otherwise, I guess we could just provide a regex-based optimized implementation when the pattern is the default. Removing the feature entirely is more involved, because it means a deprecation cycle -- and some effort to establish whether there are third parties who need it.

I don't think we can safely/securely act on potentially-arbitrary files from the filesystem (by constructing a template from them): we really need to touch them as little as possible.

**Comment by madejejej** [comment]
Just following up with fnmatch might not work. Here's why I think so:

The default pattern looks like this: `:prefix/:action{.:locale,}{.:formats,}{+:variants,}{.:handlers,}`. So you might think that we could substitute :prefix and :action and star out the rest: `:prefix/:action*` and use `fnmatch` later with the remaining details.

However, one of the tests uses this pattern `:prefix/{:formats/,}:action{.:formats,}{+:variants,}{.:handlers,}` so if we were to substitute :prefix and :action we could get an example path query like  `app/views/users/{:formats/,}show*` while in reality we'd want something like `app/views/users/html/show*`. 

So to solve this problem we'd either have to know which details we can substitute and where we can leave the star `*`. And even this won't fully help as the example in the test uses the :formats substitution twice - so we'd also have to know how many of each detail we should substitute.

**Comment by inopinatus** [comment]
I've just encountered this with 5.2.0 (again, in development only). Curiously the problem seems to affect only layouts. Views with `layout false` are fast, whilst a view with an application layout and a nested `render layout: "foo" do ...` block takes 1500ms.

**Comment by inopinatus** [comment]
This horrible jury-rigged cache gets my >1500ms problem renders back down to 30ms in development again.

```ruby
# frozen_string_literal: true

# config/initializers/monkey_dir_globcache.rb
#
# Work around glob performance woes for JS requests in development, c.f
# https://github.com/rails/rails/issues/30502, by
# prepending ActionView::PathResolver#find_template_paths to use a
# result cache.  Cache invalidation via fs listeners.
# Use at your own peril.

hacks = Module.new do
  module_function
  def dir_name_from_query(query)
    File.dirname(query.gsub(/(?<!\\){.*/, ''))
  end

  GLOB_CACHE = {}
  GLOB_CACHE_LISTENERS = Hash.new do |hash, dir|
    hash[dir] = Listen.to(dir) do
      GLOB_CACHE.delete_if do |cached_query, _|
        dir == dir_name_from_query(cached_query)
      end
    end
  end

  def find_template_paths(query)
    dirname = dir_name_from_query(query)
    return [] unless Dir.exist?(dirname)

    if GLOB_CACHE.has_key?(query)
      GLOB_CACHE[query]
    else
      GLOB_CACHE_LISTENERS[dirname].start
      GLOB_CACHE[query] = super
    end
  end
end

ActionView::PathResolver.prepend(hacks) if Rails.env.development?
```

**Comment by jrochkind** [comment]
Still having terrible problems with lots of partials slowing things down. 

Even if it's just dev-mode, it can make some pages almost unusable in dev, even in dev you don't want to have to wait 3000ms. 

I'm not totally sure it's not a problem in production too, I have seen some evidence to suggest rails may be touching the desk on every render even in production configuration, but I need to do more work to verify and prove that, could be wrong.

**Comment by feliperaul** [comment]
I believe this issue should be reopened. It's still a huge deal in Rails 5.2.2 and Ruby 2.5.3.

**Comment by feliperaul** [comment]
@sfcgeorge  Did you manage to hack rails-ujs to add format.js while this is not fixed? What is your current solution? I only stumbled on this because in a particular .js.erb file I was doing like 10 partial calls, so the latency was enormous; but since we use .js.erb a lot, I think the I/O reduction and performance gains all over our apps would be gigantic if this was fixed, even with a hack in the meanwhile.
